File: my-solutions/0163-missing-ranges/solution.java
public class Solution {
    public List<String> findMissingRanges(int[] nums, int lower, int upper) {
        List<String> result = new ArrayList<String>();
        if (nums == null || nums.length == 0) {
            outputToResult(lower, upper, result);
            return result;
        }
         
        // leading missing range
        if (nums[0] - lower > 0) {
            outputToResult(lower, nums[0] - 1, result);
        }
         
        for (int i = 1; i < nums.length; i++) {
            if (nums[i] - nums[i - 1] > 1) {
                outputToResult(nums[i - 1] + 1, nums[i] - 1, result);
            }
        }
         
        // trailing missage ranges
        if (upper - nums[nums.length - 1] > 0) {
            outputToResult(nums[nums.length - 1] + 1, upper, result);
        }
         
        return result;
    }
    
    private void outputToResult(int start, int end, List<String> result) {
        StringBuffer sb = new StringBuffer();
        if (start == end) {
            sb.append(start);
        } else {
            sb.append(start + "->" + end);
        }
         
        result.add(sb.toString());
    }
}

File: my-solutions/0163-missing-ranges/solution.py
class Solution:
    def findMissingRanges(self, nums: List[int], lower: int, upper: int) -> List[List[int]]:

        if not nums:
            return [[lower, upper]]

        # lower --- nums[0] ---- nums[-1] --- upper

        missing_ranges = []

        if lower < nums[0]:
            missing_ranges.append([lower, nums[0] - 1])

        for i in range(1, len(nums)):
            if nums[i] - nums[i - 1] > 1:
                missing_ranges.append([(nums[i - 1] + 1),nums[i] - 1])

        if nums[-1] < upper:
            missing_ranges.append([(nums[-1]+1), upper])

        return missing_ranges



        
